File: src/main/scala/org/xbery/artbeams/articles/repository/ArticleFilter.scala
package org.xbery.artbeams.articles.repository

import java.time.Instant

import org.xbery.artbeams.common.assets.repository.{AssetFilter, ValidityAssetFilter}

/**
  * @author Radek Beran
  */
case class ArticleFilter(
  override val id: Option[String],
  override val ids: Option[Seq[String]],
  override val createdBy: Option[String],
  override val validityDate: Option[Instant],
  slug: Option[String],
  showOnBlog: Option[Boolean],
  withExternalId: Option[Boolean],
  categoryId: Option[String],
  query: Option[String]
) extends AssetFilter with ValidityAssetFilter

object ArticleFilter {
  lazy val Empty = ArticleFilter(None, None, None, None, None, None, None, None, None)

  // Note: Must be def due to usage of current time
  def validOnBlog() = Empty.copy(validityDate = Some(Instant.now()), showOnBlog = Some(true))

  // Note: Must be def due to usage of current time
  def validOnBlogWithCategory(categoryId: String) = validOnBlog().copy(categoryId = Some(categoryId))

  def validByQuery(query: String) = Empty.copy(validityDate = Some(Instant.now()), query = Some(query))
}

File: src/main/kotlin/org/xbery/artbeams/articles/repository/ArticleFilter.kt
package org.xbery.artbeams.articles.repository

import org.xbery.artbeams.common.assets.repository.AssetFilter
import org.xbery.artbeams.common.assets.repository.ValidityAssetFilter
import java.time.Instant

/**
 * @author Radek Beran
 */
data class ArticleFilter(
    override val id: String?,
    override val ids: List<String>?,
    override val createdBy: String?,
    override val validityDate: Instant?,
    val slug: String?,
    val showOnBlog: Boolean?,
    val withExternalId: Boolean?,
    val categoryId: String?,
    val query: String?
) : AssetFilter, ValidityAssetFilter {
    companion object {
        val Empty: ArticleFilter = ArticleFilter(null, null, null, null, null, null, null, null, null)
        fun validOnBlog(): ArticleFilter =
            Empty.copy(validityDate = Instant.now(), showOnBlog = true)

        fun validOnBlogWithCategory(categoryId: String): ArticleFilter = validOnBlog().copy(categoryId = categoryId)
        fun validByQuery(query: String): ArticleFilter = Empty.copy(validityDate = Instant.now(), query = query)
    }
}